src/main/ruby/core/process.rb

```diff
@@ -124,11 +124,22 @@ def self.cpu_times
   # @return [Title]
   #
   def self.setproctitle(title)
-    val = Rubinius::Type.coerce_to(title, String, :to_str)
-    if !Truffle.aot? and RbConfig::CONFIG['host_os'].include?('linux') and File.readable?('/proc/self/maps')
-      setproctitle_linux_from_proc_maps(val)
+    title = Rubinius::Type.coerce_to(title, String, :to_str)
+    if Truffle.aot?
+      Truffle.invoke_primitive(:aot_set_process_title, title)
+    elsif Rubinius.linux? && File.readable?('/proc/self/maps')
+      setproctitle_linux_from_proc_maps(title)
+    elsif Rubinius.darwin?
+      # When we call _NSGetArgv we seem to always get a string that looks like what we'd expect from running ps, but
+      # with a null character inserted early. I don't know where this comes from, but it means I don't know how to get
+      # the length of space available for writing in the new program name. We therefore limit to about 40 characters,
```

Comment:
Yes, in native memory argv are simply separated by the strings null bytes.
One could read the string at the current pointer or call `strlen()` to know how long `argv[0]` is.
This can be another PR of course.

Reply:
You would need to also know `argc` to know how long is the full command line I think.